### ISSUE TITLE
Async Daemon Project Rebuild Throws an OperationCancelledException Exception

### ISSUE THREAD

**Issue by ericgreenmix** [JasperFx/marten#500]
The unit test for rebuilding (https://github.com/JasperFx/marten/blob/7d255ec5a01a7dd7f583d0fdac6c7e4baf71ff99/src/Marten.Testing/AsyncDaemon/async_daemon_end_to_end.cs#L59) gives a false positive, masking this exception. I believe this because the multiple processes are causing issues with how xunit registers errors? (Not totally sure).

If you put a break point here, https://github.com/JasperFx/marten/blob/7d255ec5a01a7dd7f583d0fdac6c7e4baf71ff99/src/Marten/Events/Projections/Async/ErrorHandling/DaemonErrorHandler.cs#L33 and debug that test, you will see this OperationCancelledException being caught.

The reason for the exception is because the `ProjectionTrack` is getting GC'ed, calling the `Dispose` method, which calls `Cancel` on the `CancellationTokenSource`. (https://github.com/JasperFx/marten/blob/7d255ec5a01a7dd7f583d0fdac6c7e4baf71ff99/src/Marten/Events/Projections/Async/ProjectionTrack.cs#L71) 
This token is still in use when it is cancelled, causing npgsql to throw and error when it calls `ThrowIfCancellationRequested` (https://github.com/npgsql/npgsql/blob/45eafe74b021679feb0d67a3007b4d3f38696a22/src/Npgsql/NpgsqlCommand.cs#L1059)

I am not entirely sure if something is causing the `ProjectionTrack` to get GC'ed when it shouldn't be, or if the `ProjectionTrack` shouldn't call `Cancel` on the `CancellationTokenSource`.

Here is the full stack trace.
```
System.OperationCanceledException: The operation was canceled.
   at System.Threading.CancellationToken.ThrowOperationCanceledException()
   at Npgsql.NpgsqlCommand.<ExecuteNonQueryAsync>d__91.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Marten.Services.UpdateBatch.<>c__DisplayClass15_0.<<ExecuteAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Marten.Services.ManagedConnection.<ExecuteAsync>d__20.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Marten.Services.UpdateBatch.<ExecuteAsync>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Marten.Services.UnitOfWork.<ApplyChangesAsync>d__22.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Marten.DocumentSession.<SaveChangesAsync>d__25.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Marten.Events.Projections.Async.ProjectionTrack.<executePage>d__43.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Marten.Events.Projections.Async.ProjectionTrack.<>c__DisplayClass42_0.<<ExecutePage>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable.ConfiguredTaskAwaiter.GetResult()
   at Marten.Events.Projections.Async.ErrorHandling.DaemonErrorHandler.<TryAction>d__4.MoveNext()
```

**Comment by jeremydmiller** [comment]
@ericgreenmix I'm not able to reproduce this one locally. How were you running the tests that spawned this one?

"I am not entirely sure if something is causing the ProjectionTrack to get GC'ed when it shouldn't be" -- I'm thinking it's that. My only theory is that the mechanics for "waiting until the rebuild is complete" aren't working correctly.

**Comment by ericgreenmix** [comment]
Okay, so when I setup the Test project on my system, I commented out the code in the `AyncDaemonFeature.cs` constructor that loads the test data from the `.json` files. Instead I just added the project manually (I did this so I didn't have to deal with the pathing issues I was having because I am running on Parallels)
```
AllProjects.Add(Guid.NewGuid(), new GithubProject("test", "test", DateTimeOffset.UtcNow));
```

I decided to resolve the pathing issues to the `.json` files, and running the test using the `.json` files works for me, without the exception. However, when I run the tests using the manually added project with the code above, it throws the `OperationCancelledException` I have been hitting.

**Comment by jeremydmiller** [comment]
Can you do a pull request w/ a failing test that shows this?

**Comment by ericgreenmix** [comment]
So unfortunately I am not having much luck getting xunit to register the error, either because it is running in another process or because the Daemon has already been GC'ed. Still trying to figure this one out...

However, I did figure out that the error occurs when there are only a single to a few events that get rebuilt. If you use only this JSON to load the projects
```
{"$type":"Marten.Testing.CodeTracker.GithubProject, Marten.Testing","Id":"d2867a24-e42b-4964-a7e3-92c9fd26a29b","OrganizationName":"DarthFubuMVC","ProjectName":"bottles","Events":{"$type":"Marten.Testing.CodeTracker.Timestamped[], Marten.Testing","$values":[{"$type":"Marten.Testing.CodeTracker.ProjectStarted, Marten.Testing","Name":"bottles","Organization":"DarthFubuMVC","Timestamp":"2011-04-12T19:50:48+00:00"}]}}
```
It will reproducibly throw an exception that you can see by debugging this test `do_a_complete_rebuild_of_the_active_projects_from_scratch_on_other_schema` and putting a breakpoint at https://github.com/JasperFx/marten/blob/7d255ec5a01a7dd7f583d0fdac6c7e4baf71ff99/src/Marten/Events/Projections/Async/ErrorHandling/DaemonErrorHandler.cs#L33

**Comment by ericgreenmix** [comment]
So I have a failing test that I just submitted a PR for now. The test shows the projection isn't being rebuilt and errors in the comparison of the `ActiveProject`, with the root cause of this being the `OperationCanceledException`.

I did a lot of digging around on this tonight and it is the `using` statement wrapping the `BuildProjectionDaemon` that is causing the dispose, because something in the rebuild isn't waiting properly (like you said above). Narrowing it down, I think it is the `ActionBlock`'s in `ProjectionTrack` -> `startConsumers` that could be the issue. I was able to get it awating correctly by replacing the `ActionBlock`'s with direct calls instead. However, I do not know the implications of doing this.
